src/conn_state.c

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright 2019 Xaptum, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "conn_state.h"
+
+#include <errno.h>
+#include <fcntl.h>
+#include <string.h>
+#include <unistd.h>
+
+#include <arpa/inet.h>
+#include <net/if.h>
+#include <netinet/in.h>
+#include <sys/socket.h>
+
+#include "context.h"
+#include "log.h"
+
+static
+int
+connect_udp_socket(char* udp_address, int udp_length, struct addrinfo* connect_addr)
+{
+    int rc = 0;
+    struct sockaddr_in udp_sock_addr;
+    int udp_sockfd = -1;
+
+    switch(connect_addr->ai_family){
+        case AF_INET:
+            udp_sockfd = socket(AF_INET, SOCK_DGRAM, 0);
+            break;
+        case AF_INET6:
+            udp_sockfd = socket(AF_INET6, SOCK_DGRAM, 0);
+            break;
+    }
+
+    if (udp_sockfd < 0) {
+        enftun_log_error("UDP: Failed to create socket: %s\n", strerror(errno));
+    }
+
+    socklen_t udp_addr_len = sizeof(struct sockaddr_in);
+
+    memset(&udp_sock_addr, 0, udp_addr_len);
+    udp_sock_addr.sin_addr.s_addr = htonl(INADDR_ANY);
+    udp_sock_addr.sin_port = htons(0);
+    udp_sock_addr.sin_family = connect_addr->ai_family;
+
+    rc = connect(udp_sockfd, connect_addr->ai_addr, connect_addr->ai_addrlen);
+    if (0 != rc){
+        close(udp_sockfd);
+        return -1;
+    }
+
+    rc = getsockname(udp_sockfd, (struct sockaddr*)&udp_sock_addr, &udp_addr_len);
+    if (0 != rc){
+        close(udp_sockfd);
+        return -1;
+    }
+
+    inet_ntop(connect_addr->ai_family, (struct sockaddr_in*)&udp_sock_addr.sin_addr, udp_address, udp_length);
+
+    close(udp_sockfd);
+
+    return 0;
+}
+
+static
+int
+udp_check(struct enftun_conn_state* conn_state)
+{
+    struct enftun_context* ctx = (struct enftun_context*)conn_state->data;
+    char udp_address[45];
+    connect_udp_socket(udp_address, sizeof(udp_address), &ctx->tls.sock.local_addr);
```

Comment:
`local_addr` is the wrong value here, I think.  `tls.sock.local_addr` is the local address (our address) of the tcp socket.  But we want the UDP socket to connect to the same *remote* address (the ENF server) as the tcp socket.

So I think `struct enftun_tcp` should be extended to have both `local_addr` and `remote_addr` fields.  Here we connect the udp socket to the tcp socket's `remote_addr`, then check if its `local_addr` is the same as the tcp socket's `local_addr`.